python/ray/data/_internal/plan.py

```diff
@@ -382,42 +389,61 @@ def schema(
+            # There remain some operators yet to be executed.
+            # Even if a schema is already previously known, it may change,
+            # so we try executing to get the most updated schema.
+
             # TODO(swang): There are several other stage types that could
             # inherit the schema or we can compute the schema without having to
             # execute any of the dataset: limit, filter, map_batches for
             # add/drop columns, etc.
             if fetch_if_missing:
-                if isinstance(self._stages_after_snapshot[-1], RandomizeBlocksStage):
+                if isinstance(self._logical_plan.dag, RandomizeBlocks):
                     # TODO(ekl): this is a hack to optimize the case where we have a
                     # trailing randomize block stages. That stage has no effect and
                     # so we don't need to execute all blocks to get the schema.
-                    a = self._stages_after_snapshot.pop()
+
+                    randomize_blocks_op = self._logical_plan.dag
+                    self._logical_plan._dag = randomize_blocks_op.input_dependencies[0]
                     try:
                         self.execute()
                     finally:
-                        self._stages_after_snapshot.append(a)
+                        self._logical_plan = randomize_blocks_op
                 else:
                     self.execute()
-            elif len(self._stages_after_snapshot) == 1 and isinstance(
-                self._stages_after_snapshot[-1], RandomizeBlocksStage
+            elif self.is_from_in_memory_only() or (
+                isinstance(self._logical_plan.dag, RandomizeBlocks)
+                and self.is_from_in_memory_only(
+                    self._logical_plan.dag.input_dependencies[0]
+                )
+            ):
+                # If InputData only, or if RandomizeBlocks is following an InputData
+                # only plan, we execute it (regardless of the fetch_if_missing),
+                # since RandomizeBlocksStage is just changing the order of references
+                # (hence super cheap).
+                self.execute()
+            elif self.is_read_only() or (
+                isinstance(self._logical_plan.dag, RandomizeBlocks)
+                and self.is_read_only(self._logical_plan.dag.input_dependencies[0])
             ):
-                # If RandomizeBlocksStage is last stage, we execute it (regardless of
-                # the fetch_if_missing), since RandomizeBlocksStage is just changing
-                # the order of references (hence super cheap).
+                # Similarly for Read only plan, or if RandomizeBlocks is following
+                # a Read only plan, we execute it (regardless of the fetch_if_missing),
+                # since RandomizeBlocksStage is just changing the order of references
+                # (hence super cheap). Calling execute does not trigger read tasks.
                 self.execute()
```

Comment:
nit, there are some conditions where we want do eagerly execute the plan. would be more readable to refactor the code to something like `if need_eager_execution(): self.execute()`